### PR TITLE
WIP: fix interaction of software requirements and environments

### PR DESCRIPTION
**Pull request by rupertnash**
A very first attempt at the fix I proposed in #1377 

This now works for a the single test case on that issue. I'd like feedback before trying to do it more properly.

Thanks!

TODO:

- [x]  document the current behaviour

### CODE REVIEW

**Review by mr-c — COMMENTED** [review]
I like this! Wasn't as bad as I expected. Guess we'll have to add testing for other types of dependency resolvers, but this should work, yeah?

**Review by mr-c — APPROVED** [review]

**Comment by mr-c** [comment]
@rupertnash Question: who wins, an environment variable set by `EnvVarRequirement` or an identical variable set by the dependency script?

I suspect that it is currently the later when it should be the former. Tricky because at this point `EnvVarRequirement` and the https://www.commonwl.org/v1.0/CommandLineTool.html#Runtime_environment variables are blended together. I guess I could add a check that favors `cwl_env`'s original values except for `HOME`, `TMPDIR`, and `PATH`. 

Speaking of, I think we should not allow `HOME` or `TMPDIR` to be overridden by the dependency script at all.

**Comment by rupertnash** [comment]
Agreed about `HOME` and `TMPDIR`, I'll add them to the exceptions

Who wins? Hard to say. We first configure the environment as normal and then allow the module to modify it, potentially overriding ones set by `EnvVarRequirement`. However this could also be a desired modification e.g. of `LD_LIBRARY_PATH`.

I see that §5.6 in the spec it says a `SoftwareRequirement` is "A list of software packages that should be configured in the environment of the defined process." This clearly allows us to modify the environment and does not place limits on this. I can see no "hierarchy" of how the environment is created so with my pedant's hat on either behaviour is acceptable.

There are multiple sources of environment now:
- Core runtime environment vars (HOME, TMPDIR, PATH)
- Variables defined by `EnvVarRequirement`
- The default environment of the container, such as when using `DockerRequirement`
- Variables set/modified by `SoftwareRequirement`s
- Ones preserved by user (`--preserve-environment=VAR`/`--preserve-entire-environment`)
I think a clear description of how these interact would help us.

On the immediate problem, I'm torn and can definitely see cases when you'll want one behaviour not the other. E.g. if you want to `--preserve-environment LD_LIBRARY_PATH` and allow a module to append its own paths to that you want the current behaviour. On the other hand you may wish to override something set by a modulefile, eg some default.

**Review by mr-c — CHANGES_REQUESTED** [review]
Okay, this needs a test and it will be good enough to merge!

**Comment by mr-c** [comment]
> Who wins?

Whatever we decide, it needs documentation, yep

**Comment by rupertnash** [comment]
TODO:
- [ ] document the current behaviour (next week)

What about this ordering? (I realise this might require some significant changes)

0. Start empty
1. Add preserved environment vars (or start with current env if entire env being preserved)
2. Set TMPDIR and HOME
3. Modify the environment in response to software requirements
4. Apply EnvVarRequirements

This will (I think) be more consistent with the way it's done for docker (i.e. the set env vars are applied last, starting from the ones set in the container)

**Comment by mr-c** [comment]
@rupertnash It is a good ordering. I think 1-3 is already happening, but should be confirmed. To implement 4+5 we'll need to split the env list into two ; might not be that bad to implement

File: README.rst [hunk 1]
```diff
@@ -292,6 +292,13 @@ This option allows one to specify a dependency resolver's configuration file.
 This file may be specified as either XML or YAML and very simply describes various
 plugins to enable to "resolve" ``SoftwareRequirement`` dependencies.
 
+Using these hints will allow cwltool to modify the environment in
+which your tool runs, for example by loading one or more environment
+modules. The enviroment is constructed as normal (i.e. standard CWL
```
Comment (mr-c) [inline README.rst:297]:
```suggestion
which your tool runs, for example by loading one or more `environment modules
<https://en.wikipedia.org/wiki/Environment_Modules_(software)>`_. The environment
is constructed as normal (i.e. standard CWL
```

File: README.rst [hunk 2]
```diff
@@ -292,6 +292,13 @@ This option allows one to specify a dependency resolver's configuration file.
...
+runtime enviroment, ``EnvVarRequirement``, ``--preserve-environment``
+are applied), then this is modified. This currently means that you
```
Comment (mr-c) [inline README.rst:299]:
```suggestion
are applied), then the environment may modified by the selected tool resolver.  This currently means that you
```

File: README.rst [hunk 3]
```diff
@@ -292,6 +292,13 @@ This option allows one to specify a dependency resolver's configuration file.
...
+cannot override variables set in this way.
```
Comment (mr-c) [inline README.rst:300]:
```suggestion
cannot override any environment variables set by the selected tool resolver.
```

File: cwltool/run_job.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
+"""Only used when there is a job script or CWLTOOL_FORCE_SHELL_POPEN=1."""
```
Comment (mr-c) [inline cwltool/run_job.py:1]: We should set this file to be excluded from test coverage

**Review by mr-c — CHANGES_REQUESTED** [review]
https://github.com/common-workflow-language/cwltool/pull/1378/files#diff-08aa5643b92eb585337d3151a28a6bb8905e482a1094f90e0dd90939d8ee4d8cR958 seems to not be covered by any tests?

Lets also mark cwltool/run_job.py as excluded from the test coverage requirements.

**Comment by rupertnash** [comment]
OK - I think I got the coverage to exclude run_job.py.

For your point about not testing that line: in `test_dependencies`, this is being hit twice iff `module` is available on the host. I don't know if this is possible to add to Travis?

I also noticed a bug in the test skip logic: the first 2 tests in that file have probably never run...

**Comment by mr-c** [comment]
> OK - I think I got the coverage to exclude run_job.py.

Confirmed, thanks!

> For your point about not testing that line: in `test_dependencies`, this is being hit twice iff `module` is available on the host. I don't know if this is possible to add to Travis?

http://modules.sourceforge.net/ ?

File: tests/test_dependencies.py [hunk 5]
```diff
@@ -35,14 +41,18 @@ def test_bioconda() -> None:
...
 
+@pytest.mark.skipif(not deps, reason="galaxy-lib is not installed")
 @pytest.mark.skipif(
     not spawn.find_executable("modulecmd"), reason="modulecmd not installed"
 )
-def test_modules() -> None:
+def test_modules(monkeypatch: Any) -> None:
+    """Here we do a basic smoke test using environment modules to
+    satisfy a SoftwareRequirement.
+    """
```
Comment (mr-c) [inline tests/test_dependencies.py:51]:
```suggestion
    """Do a basic smoke test using environment modules to satisfy a SoftwareRequirement."""
```

To fix https://travis-ci.com/github/common-workflow-language/cwltool/jobs/454295930#L4188

File: tests/test_dependencies.py [hunk 6]
```diff
@@ -55,3 +65,43 @@ def test_modules() -> None:
...
+
+@pytest.mark.skipif(not deps, reason="galaxy-lib is not installed")
+@pytest.mark.skipif(
+    not spawn.find_executable("modulecmd"), reason="modulecmd not installed"
+)
+def test_modules_environment(tmp_path: Path, monkeypatch: Any) -> None:
+    """Here check that the environment variables set by a module are
+    being propagated into a tool's environment, by running `env` as the
+    tool and parsing its output.
+    """
```
Comment (mr-c) [inline tests/test_dependencies.py:78]:
```suggestion
    """
    Check that the environment variables set by a module are being propagated correctly.

    Do so by  by running `env` as the tool and parsing its output.
    """
```

To fix https://travis-ci.com/github/common-workflow-language/cwltool/jobs/454295930#L4191

**Comment by rupertnash** [comment]
OK - having just looked at the Travis log, this one isn't my fault: it's because of the tests that never ran due to the bug I fixed in e1c3325 above

Looks like the member vars of `software_requirements.DependencyConfiguration` aren't being set up correctly.

**Comment by mr-c** [comment]
Some notes from today's CWL chat for @rupertnash 

What about `CLASSPATH` set by an `EnvVarRequirement`? I wouldn't want that to overwrite a `CLASSPATH` set by a `SoftwareRequirement`, I would want them appended. So `EnvVarRequirement` should be set before `SoftwareRequirement`s so `module load`, `conda install` can append to env vars that are lists like `CLASSPATH`

As for `HOME` and `TMPDIR`, that should be set early and set again at the end, just in case it got overwritten.

**Comment by rupertnash** [comment]
Following CWL chat just now:
0. Start empty
1. Add preserved environment vars (or start with current env if entire env being preserved)
2. Set `TMPDIR` and `HOME` as per https://www.commonwl.org/v1.2/CommandLineTool.html#Runtime_environment
3. Set `_CWLTOOL=1` (to allow modules etc be able to know when running under CWL)
3. Apply `EnvVarRequirements`
4. Modify the environment in response to software requirements
5. Ensure `HOME` is set as above, remove `_CWLTOOL`

**Comment by rupertnash** [comment]
OK - I've rebased this onto main. Going to look at the todo list now

File: README.rst [hunk 7]
```diff
@@ -298,6 +298,30 @@ CWL documents can be expressed as RDF triple graphs.
...
 
+Environment Variables in cwltool
+--------------------------------
+
+This reference implementation supports several ways of setting
+enviroment variables for tools, in addition to the standard
+``EnvVarRequirement``. The sequence of steps applied to create the
+enviroment is:
+
+0. If ``--preserve-entire-environment`` is set, begin with the current
```
Comment (mr-c) [inline README.rst:309]:
```suggestion
0. If the ``--preserve-entire-environment`` flag is present, then begin with the current
```

File: README.rst [hunk 8]
```diff
@@ -298,6 +298,30 @@ CWL documents can be expressed as RDF triple graphs.
...
+   environment, else begin with an empty environment.
+
+1. Add any variables specified by ``--preserve-environment`` option(s).
+
+2. Set ``TMPDIR`` and ``HOME`` per the CWL specification.
```
Comment (mr-c) [inline README.rst:314]:
```suggestion
2. Set ``TMPDIR`` and ``HOME`` per `the CWL v1.0+ CommandLineTool specification <https://www.commonwl.org/v1.0/CommandLineTool.html#Runtime_environment>`_.
```

File: README.rst [hunk 9]
```diff
@@ -298,6 +298,30 @@ CWL documents can be expressed as RDF triple graphs.
...
+
+3. Apply ``EnvVarRequirement`` from tool description.
```
Comment (mr-c) [inline README.rst:316]:
```suggestion
3. Apply any ``EnvVarRequirement`` from the ``CommandLineTool`` description.
```

File: README.rst [hunk 10]
```diff
@@ -298,6 +298,30 @@ CWL documents can be expressed as RDF triple graphs.
...
+
+4. Apply any manipulations required by ``MPIRequirement`` extension.
```
Comment (mr-c) [inline README.rst:318]:
```suggestion
4. Apply any manipulations required by any ``cwltool:MPIRequirement`` extensions.
```

File: README.rst [hunk 11]
```diff
@@ -298,6 +298,30 @@ CWL documents can be expressed as RDF triple graphs.
...
+
+5. Substitute any secrets required by ``Secrets`` extension.
+
+6. Modify the environment in response to ``SoftwareRequirement``.
```
Comment (mr-c) [inline README.rst:322]:
```suggestion
5. Substitute any secrets required by ``Secrets`` extension.

6. Modify the environment in response to ``SoftwareRequirement`` (see below).
```

File: cwltool/run_job.py [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        else:
+            stdout = sys.stderr
+        if stderr_path is not None:
+            stderr: Union[BinaryIO, TextIO] = open(stderr_path, "wb")
+        else:
+            stderr = sys.stderr
+        if os.name == "nt":
+            close_fds = False
+            for key, value in env.items():
+                env[key] = str(value)
```
Comment (mr-c) [inline cwltool/run_job.py:72]: This case can be dropped; see https://github.com/common-workflow-language/cwltool/pull/1423 which is about to be merged

File: cwltool/singularity.py [hunk 13]
```diff
@@ -89,6 +89,8 @@ def _normalize_sif_id(string: str) -> str:
 
 
 class SingularityCommandLineJob(ContainerCommandLineJob):
+    CONTAINER_TMPDIR: str = "/tmp"  # nosec
```
Comment (mr-c) [inline cwltool/singularity.py:92]: Should `CONTAINER_TMPDIR` move into `ContainerCommandLineJob`?

Comment (rupertnash) [reply cwltool/singularity.py:92]: maybe?

File: cwltool/run_job.py (hunk 12 above)
Comment (rupertnash) [reply cwltool/run_job.py:72]: Does this mean all the `if Windows` checks can be removed?

Comment (mr-c) [reply cwltool/run_job.py:72]: Yep! Sorry about all the pending merge conflicts..

File: cwltool/singularity.py (hunk 13 above)
Comment (rupertnash) [reply cwltool/singularity.py:92]: Yes it can; I will update

File: cwltool/job.py [hunk 14]
```diff
@@ -613,6 +617,19 @@ def add_writable_directory_volume(
     ) -> None:
         """Append a writable directory mapping to the runtime option list."""
 
+    def _preserve_environment_on_containers_warning(
+        self, varname: Optional[str] = None
+    ) -> None:
+        """When running in a container, issue a warning."""
+        if varname is None:
+            _logger.warning(
+                "You have specified `--preserve-entire-environment` while running a container"
```
Comment (mr-c) [inline cwltool/job.py:626]: We should say why this results in a warning (could hamper reproducibility / portability)

File: cwltool/singularity.py [hunk 15]
```diff
@@ -370,6 +370,12 @@ def add_writable_directory_volume(
                 self.append_volume(runtime, source, volume.target, writable=True)
                 ensure_writable(source)
 
+    def _required_env(self) -> Dict[str, str]:
+        return {
+            "TMPDIR": self.CONTAINER_TMPDIR,
+            "HOME": self.builder.outdir,
```
Comment (mr-c) [inline cwltool/singularity.py:376]: `WARNING: Overriding HOME environment variable with SINGULARITYENV_HOME is not permitted`

https://github.com/common-workflow-language/cwltool/pull/1378/checks?check_run_id=2237763816#step:9:100

```suggestion
            "SINGULARITYENV_HOME"": self.builder.outdir,
```

**Comment by rupertnash** [comment]
Ugh, need to handle env vars containing new lines by using this instead of `env`
```
import json
import os
print(json.dumps(dict(os.environ), indent=4))
```

**Comment by rupertnash** [comment]
I thought I was?

**Comment by rupertnash** [comment]
Right, that looks like it's working now!

I need to give the docs a once over.

I have run the conformance tests under Singularity (3.7) and all have passed except 173.
```
Test 173 failed: /lustre/home/dc118/nashvest/cwltool/common-workflow-language-main/cwltool_with_cov3 --parallel --singularity --outdir=/tmp/tmpg_fy7936 --quiet v1.0/docker-run-cmd.cwl v1.0/empty.json
Test Docker ENTRYPOINT usage
Returned non-zero
INFO:    Using cached SIF image
FATAL:   "-c": executable file not found in $PATH
ERROR [job docker-run-cmd.cwl] Job error:
("Error collecting output for parameter 'cow':\nv1.0/docker-run-cmd.cwl:14:7: Did not find output file with glob pattern: '['cow']'", {})
WARNING [job docker-run-cmd.cwl] completed permanentFail
WARNING Final process status is permanentFail

196 tests passed, 1 failures, 0 unsupported features
```

This test is problematic for Singularity. The test runs the tool "v1.0/docker-run-cmd.cwl", who's definition is:
```yaml
class: CommandLineTool
cwlVersion: v1.0
hints:
  ResourceRequirement:
    ramMin: 8
requirements:
  DockerRequirement:
    dockerPull: bash:4.4.12
inputs: []
outputs:
  cow:
    type: File
    outputBinding:
      glob: cow
baseCommand: ["-c", "echo 'moo' > cow"]
```

Singularity has two variants 
- `singularity exec [<exec options>] <container> <command> [<args>]` to run a user-specified command in a container, and
- `singularity run [<options>] <container> [<args>]` to run the image's default command in a container

While docker is, according to docs `docker run image [command] [args]`  but this isn't the whole story because `ENTRYPOINT` and `CMD` mess around with this, both of which are set in this container to make it act a lot like just executing `bash` on your terminal. 

Reading the CWL spec (https://www.commonwl.org/v1.0/CommandLineTool.html#DockerRequirement):

> When running a tool contained in Docker, the workflow platform must not assume anything about the contents of the Docker container, such as the presence or absence of specific software, except to assume that the *generated command line represents a valid command within the runtime environment of the container*.

I conclude that the behaviour of this test is contrary to the wording of the standard since `["-c", "echo 'moo' > cow"]` is not a valid *command* in either shell or docker language.

**Comment by mr-c** [comment]
> Right, that looks like it's working now!

Huzzah!

> I have run the conformance tests under Singularity (3.7) and all have passed except 173.

Yeah, this is the one that never passed for me. Here's the discussion I had with the Singularity devs about it https://github.com/hpcng/singularity/issues/3673#issuecomment-517691733

In short: can you restore the running of the conformance tests with Singularity, but skipping test 173? And then we can open another issue to follow this further.

Then I can do another review here and hopefully merge 😄

**Comment by rupertnash** [comment]
> In short: can you restore the running of the conformance tests with Singularity, but skipping test 173? 
Yes, just trying to figure our how to skip tests with `cwltest` - any clues?

**Comment by mr-c** [comment]
> > In short: can you restore the running of the conformance tests with Singularity, but skipping test 173? 
> 
> Yes, just trying to figure our how to skip tests with `cwltest` - any clues?
> 

`cwltest -n1-172,174-${last_test_number}` or fix https://github.com/common-workflow-language/cwltest/issues/111 😄

**Comment by tetron** [comment]
v1.0 didn't specify what to do in the presence of ENTRYPOINT.  It was addressed in v1.1:

https://www.commonwl.org/v1.1/CommandLineTool.html#DockerRequirement

> A container image may specify an ENTRYPOINT and/or CMD. Command line arguments will be appended after all elements of ENTRYPOINT, and will override all elements specified using CMD (in other words, CMD is only used when the CommandLineTool definition produces an empty command line).

Unfortunately this is a situation where we were codifying existing de-facto behavior.

How does does the behavior of `singularity run` differ from Docker ENTRYPONT?

**Comment by rupertnash** [comment]
Not significantly

**Comment by tetron** [comment]
> Not significantly

So in that case, the correct behavior is probably to use `singularity run` not `singularity exec` ?  Is it possible to have an null default script in which case the fallback case for `run` is the same as `exec`?

File: conformance-test.sh [hunk 16]
```diff
@@ -6,36 +6,33 @@ venv() {
...
 	pip3 install -e .
-	pip3 install "cwltest>=1.0.20180518074130" codecov
-	if [[ "$version" = "v1.0" ]]
-	then
-		DRAFT="DRAFT=v1.0"
-		pushd common-workflow-language-main || exit 1
-	else
-		pushd cwl-"${repo}"-main || exit 1
+	pip3 install codecov
+	pip3 install git+https://github.com/common-workflow-language/cwltest.git@main#cwltest
```
Comment (mr-c) [inline conformance-test.sh:31]: `cwltest` version 2.1 is out with your changes now :-) https://pypi.org/project/cwltest/2.1.20210423130452/

**Comment by rupertnash** [comment]
@mr-c the coverage part of conformance-test.sh is a bit shonky. I've mostly fixed this running locally but what about the upload part? This errors out with:
```
+ codecov --file coverage.xml

      _____          _
     / ____|        | |
    | |     ___   __| | ___  ___ _____   __
    | |    / _ \ / _  |/ _ \/ __/ _ \ \ / /
    | |___| (_) | (_| |  __/ (_| (_) \ V /
     \_____\___/ \____|\___|\___\___/ \_/
                                    v2.1.11

==> Detecting CI provider
  -> Got branch from git/hg
  -> Got sha from git/hg
==> Preparing upload
    Error running `['hg', 'locate']`: returncode=None, output=[Errno 2] No such file or directory: 'hg'
    Error running `['hg', 'locate']`: returncode=None, output=[Errno 2] No such file or directory: 'hg'
==> Processing gcov (disable by -X gcov)
==> Collecting reports
    Targeting specific files
    + coverage.xml bytes=407851
==> Uploading
    .url https://codecov.io
    .query commit=a76fed55b5f9fb726b1a2c8144c656743849e160&branch=deps-env&package=py2.1.11
    Gzipping contents..
    Compressed contents to 29094 bytes
    Pinging Codecov...
Error: Could not determine repo and owner
```

File: .github/workflows/ci-tests.yml [hunk 17]
```diff
@@ -141,6 +146,7 @@ jobs:
       - name: Run CWL conformance tests ${{ matrix.cwl-version }}
         env:
           version: ${{ matrix.cwl-version }}
+          spec_branch: exclude
```
Comment (mr-c) [inline .github/workflows/ci-tests.yml:149]: This branch was merged into `main`
```suggestion
```

Comment (rupertnash) [reply .github/workflows/ci-tests.yml:149]: Thanks - I'll do PRs for the other spec repos then can replace this.

Comment (rupertnash) [reply .github/workflows/ci-tests.yml]: done

**Comment by rupertnash** [comment]
OK, I'm very confused about the failing test in 1.1 (199). It runs `wc` on whale.txt but the output mismatches.

If I run a shell in the container via singularity we can see different behaviour:
```
Singularity> wc cwl-v1.1-main/tests/whale.txt 
  16  198 1111 cwl-v1.1-main/tests/whale.txt
Singularity> wc < cwl-v1.1-main/tests/whale.txt 
  16  198 1111
Singularity> cat cwl-v1.1-main/tests/whale.txt | wc
     16     198    1111
Singularity> cat cwl-v1.1-main/tests/whale.txt | wc | sha1sum 
2936abd60952276ffbf91d709b19c6c48c263947  -
```
The last gives the expected answer (and the one that docker gives).

There must be some difference in the way singularity is passing file handles to the process in the container vs docker.

**Comment by rupertnash** [comment]
I'm just going to disable for now, see #1440

**Comment by tetron** [comment]
@rupertnash argh, this has come up before, the `wc` program differs slightly in its output, either across implementations, or depending on whether you're writing to a PTY or a file.  I there's a longstanding bug report to do something about that.

**Review by mr-c — COMMENTED** [review]
Just one final Singularity + CWL v1.2 test failure: "Test directory literal containing a real file"

https://github.com/common-workflow-language/cwltool/pull/1378/checks?check_run_id=2421911722#step:6:880

I think it is unrelated and can be moved to another issue

**Review by mr-c — APPROVED** [review]

**Comment by mr-c** [comment]
@rupertnash Do you want to clean up the git history for a simple rebase onto `main` ; or shall I squash down to a single mega-commit?

**Comment by rupertnash** [comment]
> @rupertnash Do you want to clean up the git history for a simple rebase onto `main` ; or shall I squash down to a single mega-commit?

I am not a bit fan of mega commits. I've streamlined the history in what I just forced (and rebased onto main). Please merge if you're happy!!

**Pull request merged**